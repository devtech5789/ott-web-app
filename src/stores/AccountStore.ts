import type { PaymentDetail, Subscription, Transaction } from '#types/subscription';
import type { Consent, Customer, CustomerConsent } from '#types/account';
import { createStore } from '#src/stores/utils';
import type { Offer } from '#types/checkout';

type AccountStore = {
  loading: boolean;
  user: Customer | null;
  subscription: Subscription | null;
  transactions: Transaction[] | null;
  activePayment: PaymentDetail | null;
  customerConsents: CustomerConsent[] | null;
  publisherConsents: Consent[] | null;
  pendingOffer: Offer | null;
  canUpdateEmail: boolean;
  canRenewSubscription: boolean;
  canUpdatePaymentMethod: boolean;
  canChangePasswordWithOldPassword: boolean;
  canExportAccountData: boolean;
  canDeleteAccount: boolean;
  canShowReceipts: boolean;
  setLoading: (loading: boolean) => void;
};

export const useAccountStore = createStore<AccountStore>('AccountStore', (set) => ({
  loading: true,
  user: null,
  subscription: null,
  transactions: null,
  activePayment: null,
  customerConsents: null,
  publisherConsents: null,
  pendingOffer: null,
  canUpdateEmail: false,
  canRenewSubscription: false,
  canChangePasswordWithOldPassword: false,
  canExportAccountData: false,
  canDeleteAccount: false,
  canUpdatePaymentMethod: false,
  canShowReceipts: false,
  setLoading: (loading: boolean) => set({ loading }),
}));
