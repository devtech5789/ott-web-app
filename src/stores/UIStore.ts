import type { Location } from 'react-router-dom';

import { createStore } from './utils';

type UIState = {
  searchQuery: string;
  searchActive: boolean;
  userMenuOpen: boolean;
  languageMenuOpen: boolean;
  preSearchPage?: Location;
};

export const useUIStore = createStore<UIState>('UIStore', () => ({
  searchQuery: '',
  searchActive: false,
  languageMenuOpen: false,
  userMenuOpen: false,
}));
