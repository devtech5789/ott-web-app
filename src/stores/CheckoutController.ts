import { getOverrideIP } from '../utils/common';

import { subscribeToNotifications } from './NotificationsController';
import { useAccountStore } from './AccountStore';
import { reloadActiveSubscription } from './AccountController';

import type {
  AddAdyenPaymentDetailsResponse,
  AdyenPaymentSession,
  CardPaymentData,
  CreateOrderArgs,
  FinalizeAdyenPaymentDetails,
  FinalizeAdyenPayment,
  InitialAdyenPayment,
  Offer,
  Order,
  PaymentMethod,
  PaymentWithPayPalResponse,
  UpdateOrderPayload,
  SwitchOffer,
} from '#types/checkout';
import { useCheckoutStore } from '#src/stores/CheckoutStore';
import useAccount from '#src/hooks/useAccount';
import useService from '#src/hooks/useService';

export const createOrder = async (offer: Offer, paymentMethodId?: number): Promise<unknown> => {
  return await useAccount(async ({ customer }) => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');

      const createOrderArgs: CreateOrderArgs = {
        offer,
        customerId: customer.id,
        country: customer?.country || '',
        customerIP: customer?.lastUserIp || '',
        paymentMethodId,
      };

      const response = await checkoutService.createOrder(createOrderArgs, sandbox);

      if (response?.errors?.length > 0) {
        useCheckoutStore.getState().setOrder(null);

        throw new Error(response?.errors[0]);
      }

      useCheckoutStore.getState().setOrder(response.responseData?.order);
    });
  });
};

export const updateOrder = async (order: Order, paymentMethodId?: number, couponCode?: string | null): Promise<unknown> => {
  return await useAccount(async () => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');

      const updateOrderPayload: UpdateOrderPayload = {
        order,
        paymentMethodId,
        couponCode,
      };

      const response = await checkoutService.updateOrder(updateOrderPayload, sandbox);
      if (response.errors.length > 0) {
        // clear the order when the order doesn't exist on the server
        if (response.errors[0].includes(`Order with ${order.id} not found`)) {
          useCheckoutStore.getState().setOrder(null);
        }

        throw new Error(response.errors[0]);
      }

      if (response.responseData?.order) {
        useCheckoutStore.getState().setOrder(response.responseData?.order);
      }
    });
  });
};

export const getPaymentMethods = async (): Promise<PaymentMethod[]> => {
  return await useAccount(async () => {
    return await useService(async ({ checkoutService, sandbox = true }) => {
      if (!checkoutService) throw new Error('checkout service is not available');

      const { paymentMethods } = useCheckoutStore.getState();

      if (paymentMethods) return paymentMethods; // already fetched payment methods

      const response = await checkoutService.getPaymentMethods(sandbox);

      if (response.errors.length > 0) throw new Error(response.errors[0]);

      useCheckoutStore.getState().setPaymentMethods(response.responseData?.paymentMethods);

      return response.responseData?.paymentMethods;
    });
  });
};

export const paymentWithoutDetails = async (): Promise<unknown> => {
  return await useAccount(async () => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      if (!checkoutService) throw new Error('checkout service is not available');

      const { order } = useCheckoutStore.getState();

      if (!order) throw new Error('No order created');
      if (!authProviderId) throw new Error('auth provider is not configured');

      const response = await checkoutService.paymentWithoutDetails({ orderId: order.id }, sandbox);

      if (response.errors.length > 0) throw new Error(response.errors[0]);
      if (response.responseData.rejectedReason) throw new Error(response.responseData.rejectedReason);

      return response.responseData;
    });
  });
};

export const directPostCardPayment = async (cardPaymentPayload: CardPaymentData): Promise<unknown> => {
  return await useAccount(async ({ customer }) => {
    return await useService(async ({ checkoutService, authProviderId }) => {
      const { order } = useCheckoutStore.getState();

      if (!order) throw new Error('No order created');
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');

      // subscribe to listen to inplayer websocket notifications
      await subscribeToNotifications(customer?.uuid);

      return await checkoutService.directPostCardPayment(cardPaymentPayload, order);
    });
  });
};

export const createAdyenPaymentSession = async (returnUrl: string, isInitialPayment: boolean = true): Promise<AdyenPaymentSession> => {
  return await useAccount(async () => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      const { order } = useCheckoutStore.getState();

      const orderId = order?.id;

      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');
      if (isInitialPayment && !orderId) throw new Error('There is no order to pay for');
      if (!('createAdyenPaymentSession' in checkoutService)) throw new Error('createAdyenPaymentSession is not available in checkout service');

      const response = await checkoutService.createAdyenPaymentSession(
        {
          orderId: orderId,
          returnUrl: returnUrl,
        },
        sandbox,
      );

      if (response.errors.length > 0) throw new Error(response.errors[0]);

      return response.responseData;
    });
  });
};

export const initialAdyenPayment = async (paymentMethod: AdyenPaymentMethod, returnUrl: string): Promise<InitialAdyenPayment> => {
  return await useAccount(async () => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      const { order } = useCheckoutStore.getState();

      if (!order) throw new Error('No order created');
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');
      if (!('initialAdyenPayment' in checkoutService)) throw new Error('initialAdyenPayment is not available in checkout service');

      const response = await checkoutService.initialAdyenPayment(
        {
          orderId: order.id,
          returnUrl: returnUrl,
          paymentMethod,
          attemptAuthentication: sandbox ? 'always' : undefined,
          customerIP: getOverrideIP(),
        },
        sandbox,
      );

      if (response.errors.length > 0) throw new Error(response.errors[0]);

      return response.responseData;
    });
  });
};

export const finalizeAdyenPayment = async (details: unknown, orderId?: number, paymentData?: string): Promise<FinalizeAdyenPayment> => {
  return await useAccount(async () => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      if (!orderId) throw new Error('No order created');
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');
      if (!('initialAdyenPayment' in checkoutService)) throw new Error('finalizeAdyenPayment is not available in checkout service');

      const response = await checkoutService.finalizeAdyenPayment(
        {
          orderId,
          details,
          paymentData,
        },
        sandbox,
      );

      if (response.errors.length > 0) throw new Error(response.errors[0]);

      return response.responseData;
    });
  });
};

export const paypalPayment = async (successUrl: string, cancelUrl: string, errorUrl: string, couponCode: string = ''): Promise<PaymentWithPayPalResponse> => {
  return await useAccount(async () => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      const { order } = useCheckoutStore.getState();

      if (!order) throw new Error('No order created');
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');

      const response = await checkoutService.paymentWithPayPal(
        {
          order: order,
          successUrl,
          cancelUrl,
          errorUrl,
          couponCode,
        },
        sandbox,
      );

      if (response.errors.length > 0) throw new Error(response.errors[0]);

      return response.responseData;
    });
  });
};

export const getSubscriptionSwitches = async (): Promise<unknown> => {
  return await useAccount(async ({ customerId }) => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');
      if (!('getSubscriptionSwitches' in checkoutService)) return;
      const { subscription } = useAccountStore.getState();

      if (!subscription) return;

      const response = await checkoutService.getSubscriptionSwitches(
        {
          customerId: customerId,
          offerId: subscription.offerId,
        },
        sandbox,
      );

      if (!response.responseData.available.length) return;

      const switchOffers = response.responseData.available.map((offer: SwitchOffer) => checkoutService.getOffer({ offerId: offer.toOfferId }, sandbox));
      const offers = await Promise.all(switchOffers);

      // Sort offers for proper ordering in "Choose Offer" modal when applicable
      const offerSwitches = offers.sort((a, b) => a?.responseData.offerPrice - b?.responseData.offerPrice).map((item) => item.responseData);
      useCheckoutStore.setState({ offerSwitches });
    });
  });
};

export const switchSubscription = async (toOfferId: string, switchDirection: 'upgrade' | 'downgrade'): Promise<unknown> => {
  return await useAccount(async ({ customerId }) => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');
      if (!('switchSubscription' in checkoutService)) throw new Error('SwitchSubscription not supported');
      const { subscription } = useAccountStore.getState();

      if (!subscription) return;

      const SwitchSubscriptionPayload = { toOfferId, customerId: customerId, offerId: subscription.offerId, switchDirection: switchDirection };

      await checkoutService.switchSubscription(SwitchSubscriptionPayload, sandbox);

      // switching a subscription takes a bit longer to process
      await reloadActiveSubscription({ delay: 7500 });

      // clear current offers
      useCheckoutStore.setState({ offerSwitches: [] });
    });
  });
};

export const changeSubscription = async ({ accessFeeId, subscriptionId }: { accessFeeId: string; subscriptionId: string }) => {
  return await useService(async ({ subscriptionService, sandbox = true }) => {
    if (!subscriptionService || !('changeSubscription' in subscriptionService)) throw new Error('subscription service is not configured');

    const { responseData } = await subscriptionService.changeSubscription({ accessFeeId, subscriptionId }, sandbox);

    return responseData;
  });
};

export const updatePayPalPaymentMethod = async (
  successUrl: string,
  cancelUrl: string,
  errorUrl: string,
  paymentMethodId: number,
  currentPaymentId?: number,
) => {
  return await useAccount(async () => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');
      if (!('updatePaymentMethodWithPayPal' in checkoutService)) throw new Error('updatePaymentMethodWithPayPal is not available in checkout service');

      const response = await checkoutService.updatePaymentMethodWithPayPal(
        {
          paymentMethodId,
          successUrl,
          cancelUrl,
          errorUrl,
        },
        sandbox,
      );

      if (response.errors.length > 0) throw new Error(response.errors[0]);

      if (currentPaymentId) {
        await checkoutService.deletePaymentMethod({ paymentDetailsId: currentPaymentId }, sandbox);
      }

      return response.responseData;
    });
  });
};

export const addAdyenPaymentDetails = async (
  paymentMethod: AdyenPaymentMethod,
  paymentMethodId: number,
  returnUrl: string,
): Promise<AddAdyenPaymentDetailsResponse> => {
  return await useAccount(async () => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');
      if (!('addAdyenPaymentDetails' in checkoutService)) throw new Error('addAdyenPaymentDetails is not available in checkout service');

      const response = await checkoutService.addAdyenPaymentDetails(
        {
          paymentMethodId,
          returnUrl,
          paymentMethod,
          attemptAuthentication: sandbox ? 'always' : undefined,
          customerIP: getOverrideIP(),
        },
        sandbox,
      );

      if (response.errors.length > 0) throw new Error(response.errors[0]);

      return response.responseData;
    });
  });
};

export const finalizeAdyenPaymentDetails = async (details: unknown, paymentMethodId: number, paymentData?: string): Promise<FinalizeAdyenPaymentDetails> => {
  return await useAccount(async () => {
    return await useService(async ({ checkoutService, sandbox = true, authProviderId }) => {
      if (!authProviderId) throw new Error('auth provider is not configured');
      if (!checkoutService) throw new Error('checkout service is not available');
      if (!('initialAdyenPayment' in checkoutService)) throw new Error('finalizeAddedAdyenPaymentDetails is not available in checkout service');

      const response = await checkoutService.finalizeAdyenPaymentDetails({ paymentMethodId, details, paymentData }, sandbox);

      if (response.errors.length > 0) throw new Error(response.errors[0]);

      return response.responseData;
    });
  });
};
