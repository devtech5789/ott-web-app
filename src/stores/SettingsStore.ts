import { createStore } from '#src/stores/utils';
import { OTT_GLOBAL_PLAYER_ID } from '#src/config';

export interface Settings {
  defaultConfigSource?: string;
  playerId: string;
  playerLicenseKey?: string;
  additionalAllowedConfigSources?: string[];
  UNSAFE_allowAnyConfigSource?: boolean;
}

export const useSettingsStore = createStore<Settings>('SettingsStore', () => ({
  additionalAllowedConfigSources: [],
  playerId: OTT_GLOBAL_PLAYER_ID,
}));
