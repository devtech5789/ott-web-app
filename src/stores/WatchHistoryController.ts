import { useAccountStore } from '#src/stores/AccountStore';
import * as persist from '#src/utils/persist';
import { getMediaItems, updatePersonalShelves } from '#src/stores/AccountController';
import { useConfigStore } from '#src/stores/ConfigStore';
import { useWatchHistoryStore } from '#src/stores/WatchHistoryStore';
import type { PlaylistItem } from '#types/playlist';
import type { SerializedWatchHistoryItem, WatchHistoryItem } from '#types/watchHistory';
import { MAX_WATCHLIST_ITEMS_COUNT } from '#src/config';
import { getSeriesByMediaIds } from '#src/services/api.service';

const PERSIST_KEY_WATCH_HISTORY = `history${window.configId ? `-${window.configId}` : ''}`;

// Retrieve watch history media items info using a provided watch list
const getWatchHistoryItems = async (continueWatchingList: string, ids: string[]): Promise<Record<string, PlaylistItem>> => {
  const watchHistoryItems = await getMediaItems(continueWatchingList, ids);
  const watchHistoryItemsDict = Object.fromEntries((watchHistoryItems || []).map((item) => [item.mediaid, item]));

  return watchHistoryItemsDict;
};

// We store separate episodes in the watch history and to show series card in the Continue Watching shelf we need to get their parent media items
const getWatchHistorySeriesItems = async (continueWatchingList: string, ids: string[]): Promise<Record<string, PlaylistItem | undefined>> => {
  const mediaWithSeries = await getSeriesByMediaIds(ids);
  const seriesIds = Object.keys(mediaWithSeries || {})
    .map((key) => mediaWithSeries?.[key]?.[0]?.series_id)
    .filter(Boolean) as string[];

  const seriesItems = await getMediaItems(continueWatchingList, seriesIds);
  const seriesItemsDict = Object.keys(mediaWithSeries || {}).reduce((acc, key) => {
    const seriesItemId = mediaWithSeries?.[key]?.[0]?.series_id;
    if (seriesItemId) {
      acc[key] = seriesItems?.find((el) => el.mediaid === seriesItemId);
    }
    return acc;
  }, {} as Record<string, PlaylistItem | undefined>);

  return seriesItemsDict;
};

export const restoreWatchHistory = async () => {
  const { user } = useAccountStore.getState();
  const continueWatchingList = useConfigStore.getState().config.features?.continueWatchingList;

  const savedItems = user ? user.externalData?.history : persist.getItem<WatchHistoryItem[]>(PERSIST_KEY_WATCH_HISTORY);

  if (savedItems?.length && continueWatchingList) {
    // When item is an episode of the new flow -> show the card as a series one, but keep episode to redirect in a right way
    const ids = savedItems.map(({ mediaid }) => mediaid);

    const watchHistoryItems = await getWatchHistoryItems(continueWatchingList, ids);
    const seriesItems = await getWatchHistorySeriesItems(continueWatchingList, ids);

    const watchHistory = savedItems.map((item) => {
      const parentSeries = seriesItems?.[item.mediaid];
      const historyItem = watchHistoryItems[item.mediaid];

      if (historyItem) {
        return createWatchHistoryItem(parentSeries || historyItem, item.mediaid, parentSeries?.mediaid, item.progress);
      }
    });

    useWatchHistoryStore.setState({
      watchHistory: watchHistory.filter((item): item is WatchHistoryItem => !!item?.mediaid),
      playlistItemsLoaded: true,
      continueWatchingPlaylistId: continueWatchingList,
    });
  }
};

export const serializeWatchHistory = (watchHistory: WatchHistoryItem[]): SerializedWatchHistoryItem[] =>
  watchHistory.map(({ mediaid, progress }) => ({
    mediaid,
    progress,
  }));

export const persistWatchHistory = () => {
  const { watchHistory } = useWatchHistoryStore.getState();
  const { user } = useAccountStore.getState();

  if (user) {
    return updatePersonalShelves();
  }

  persist.setItem(PERSIST_KEY_WATCH_HISTORY, serializeWatchHistory(watchHistory));
};

/** Use mediaid of originally watched movie / episode.
 * A playlistItem can be either a series item (to show series card) or media item
 * */
export const createWatchHistoryItem = (item: PlaylistItem, mediaid: string, seriesId: string | undefined, videoProgress: number): WatchHistoryItem => {
  return {
    mediaid,
    seriesId,
    title: item.title,
    tags: item.tags,
    duration: item.duration,
    progress: videoProgress,
    playlistItem: item,
  } as WatchHistoryItem;
};

/**
 *  If we already have an element with continue watching state, we:
 *    1. Update the progress
 *    2. Move the element to the continue watching list start
 *  Otherwise:
 *    1. Move the element to the continue watching list start
 *    2. If there are many elements in continue watching state we remove the oldest one
 */
export const saveItem = async (item: PlaylistItem, seriesItem: PlaylistItem | undefined, videoProgress: number | null) => {
  const { watchHistory } = useWatchHistoryStore.getState();

  if (!videoProgress) return;

  const watchHistoryItem = createWatchHistoryItem(seriesItem || item, item.mediaid, seriesItem?.mediaid, videoProgress);
  // filter out the existing watch history item, so we can add it to the beginning of the list
  const updatedHistory = watchHistory.filter(({ mediaid, seriesId }) => {
    return mediaid !== watchHistoryItem.mediaid && (!seriesId || seriesId !== watchHistoryItem.seriesId);
  });

  updatedHistory.unshift(watchHistoryItem);

  updatedHistory.splice(MAX_WATCHLIST_ITEMS_COUNT);
  useWatchHistoryStore.setState({ watchHistory: updatedHistory });

  await persistWatchHistory();
};
