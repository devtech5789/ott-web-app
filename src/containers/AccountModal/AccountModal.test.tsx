import React from 'react';

import AccountModal from './AccountModal';

import { renderWithRouter } from '#test/testUtils';
import * as checkoutController from '#src/stores/CheckoutController';

describe('<AccountModal>', () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  test('renders and matches snapshot', () => {
    const spy = vi.spyOn(checkoutController, 'getSubscriptionSwitches');
    spy.mockResolvedValue(undefined);

    const { container } = renderWithRouter(<AccountModal />);

    expect(container).toMatchSnapshot();
  });
});
