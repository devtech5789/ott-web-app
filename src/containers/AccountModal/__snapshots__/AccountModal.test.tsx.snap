// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<AccountModal> > renders and matches snapshot 1`] = `<div />`;
